core/src/main/java/org/apache/druid/java/util/http/client/pool/ResourcePool.java

```diff
@@ -212,15 +259,21 @@ V get()
         if (closed) {
           log.info(StringUtils.format("get() called even though I'm closed. key[%s]", key));
           return null;
-        } else if (!resourceHolderList.isEmpty()) {
-          ResourceHolder<V> holder = resourceHolderList.removeFirst();
-          if (System.currentTimeMillis() - holder.getLastAccessedTime() > unusedResourceTimeoutMillis) {
-            factory.close(holder.getResource());
+        } else if (numLentResources + deficit < maxSize) {
+          // Attempt to take an existing resource or create one if list is empty
+          if (resourceHolderList.isEmpty()) {
             poolVal = factory.generate(key);
           } else {
-            poolVal = holder.getResource();
+            ResourceHolder<V> holder = resourceHolderList.removeFirst();
+            if (System.currentTimeMillis() - holder.getLastAccessedTime() > unusedResourceTimeoutMillis) {
+              factory.close(holder.getResource());
```

Comment:
hmm. what happens if there is an exception here. we would have removed an item from the list but the item would not have been handed over to the consumer. so our pool effectively shrinks by 1. This is an existing behaviour though. 
am I thinking it wrong?

Reply:
deficit should be incremented here too if there is an exception in regenerating the source. In fact the whole thing can be moved outside synchronized block.

Reply:
deficit has been eliminated